src/Project.php

```diff
@@ -89,86 +61,14 @@ public function getDocument(string $uri)
     }
 
     /**
-     * Returns the document indicated by uri.
-     * If the document is not open, loads it.
-     *
-     * @param string $uri
-     * @return Promise <PhpDocument>
-     */
-    public function getOrLoadDocument(string $uri)
-    {
-        return isset($this->documents[$uri]) ? Promise\resolve($this->documents[$uri]) : $this->loadDocument($uri);
-    }
-
-    /**
-     * Loads the document by doing a textDocument/xcontent request to the client.
-     * If the client does not support textDocument/xcontent, tries to read the file from the file system.
-     * The document is NOT added to the list of open documents, but definitions are registered.
+     * Ensures a document is added to the list of open documents.
      *
-     * @param string $uri
-     * @return Promise <PhpDocument>
-     */
-    public function loadDocument(string $uri): Promise
-    {
-        return coroutine(function () use ($uri) {
-            $limit = 150000;
-            if ($this->clientCapabilities->xcontentProvider) {
-                $content = (yield $this->client->textDocument->xcontent(new TextDocumentIdentifier($uri)))->text;
-                $size = strlen($content);
-                if ($size > $limit) {
-                    throw new ContentTooLargeException($uri, $size, $limit);
-                }
-            } else {
-                $path = uriToPath($uri);
-                $size = filesize($path);
-                if ($size > $limit) {
-                    throw new ContentTooLargeException($uri, $size, $limit);
-                }
-                $content = file_get_contents($path);
-            }
-            if (isset($this->documents[$uri])) {
-                $document = $this->documents[$uri];
-                $document->updateContent($content);
-            } else {
-                $document = new PhpDocument(
-                    $uri,
-                    $content,
-                    $this,
-                    $this->client,
-                    $this->parser,
-                    $this->docBlockFactory,
-                    $this->definitionResolver
-                );
-            }
-            return $document;
-        });
-    }
-
-    /**
-     * Ensures a document is loaded and added to the list of open documents.
-     *
-     * @param string $uri
-     * @param string $content
+     * @param PhpDocument $document
      * @return void
      */
-    public function openDocument(string $uri, string $content)
+    public function addDocument(PhpDocument $document)
```

Comment:
This method could register all the definitions of the document to the index, instead of the factory. Would be one more iteration though.